src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/ReflectionMarker.cs

```diff
@@ -215,11 +219,11 @@ internal void CheckAndWarnOnReflectionAccess(in MessageOrigin origin, TypeSystem
                 }
                 else
                 {
-                    ReportRequires(origin, entity, DiagnosticUtilities.RequiresUnreferencedCodeAttribute);
+                    ReportRequires(origin, entity, DiagnosticUtilities.RequiresUnreferencedCodeAttribute, requiresAttribute.Value);
                 }
             }
 
-            if (entity.DoesMemberRequire(DiagnosticUtilities.RequiresAssemblyFilesAttribute, out _))
+            if (_logger.ShouldSuppressAnalysisWarningsForRequires(entity, DiagnosticUtilities.RequiresAssemblyFilesAttribute, out requiresAttribute))
```

Comment:
The reasoning is right below Michal's comment: https://github.com/dotnet/linker/issues/2739#issuecomment-1098026715

My main reason is: the compiler should be effectively bullet-proof, it either warns, or the thing works as before. Now we're never going to be there 100%, but we try (with only a few exceptions where we decided it's not worth the trouble). In this case, it's really easy to implement this in the compiler. We have to accept all of the related complexity because we need to support RUC, and thus supporting RDC/RAF is basically for free.